Enable test_api IMethodTest in OSS

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62950

Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated torch::deploy examples when building torch_deploy library.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

Differential Revision: [D30334815](https://our.internmc.facebook.com/intern/diff/D30334815)

Commit:
Enable test_api IMethodTest in OSS

Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated torch::deploy examples when building torch_deploy library.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

[ghstack-poisoned]

wconstab commented on test/cpp/api/imethod.cpp:
```diff
@@ -8,38 +8,46 @@
 using namespace caffe2;
 
-// TODO(T96218435): Enable the following tests in OSS.
+const char* simple = "torch/csrc/deploy/example/generated/simple";
```
this path is OK?  I thought that maybe the reason the previous PR failed was that this path was not right given where it was run from.  Maybe it is OK though, if the test is run from the root...

alanwaketan replied:
Probably okay. That's what the test_deploy.cpp uses. I bet the tests are run from the root.

wconstab commented:
Did you root cause the previous issue and verify its fixed now?

alanwaketan commented:
> Did you root cause the previous issue and verify its fixed now?

No, I didn't. I cannot reproduce it locally (on AWS GPU Cluster). I was hoping some of the CI here can examine it. It looks like there is a CI failure. Let me dig into it.

alanwaketan commented on torch/csrc/deploy/CMakeLists.txt:
```diff
@@ -46,3 +46,10 @@ target_include_directories(test_deploy_lib BEFORE PRIVATE ${PYTHON_INC_DIR})
 add_executable(deploy_benchmark ${DEPLOY_DIR}/example/benchmark.cpp)
 target_include_directories(deploy_benchmark PRIVATE ${PYTORCH_ROOT}/torch)
 target_link_libraries(deploy_benchmark PUBLIC torch_deploy)
+
```
From the CI bots, it looks like this step has to be run after pytorch is installed to the system. Otherwise, `import torch` will fail. Trying to figure out where could be the right phase to run this script.

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30334815).

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated torch::deploy examples when building torch_deploy library.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

Differential Revision: [D30334815](https://our.internmc.facebook.com/intern/diff/D30334815)

[ghstack-poisoned]

alanwaketan commented:
@alanwaketan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30334815).

Commit:
Update on "Enable test_api IMethodTest in OSS"


Summary:
This diff did the following few things to enable the tests:
1. Exposed IMethod as TORCH_API.
2. Linked torch_deploy to test_api if USE_DEPLOY == 1.
3. Generated torch::deploy examples when building torch_deploy library.

Test Plan:
./build/bin/test_api --gtest_filter=IMethodTest.*

Differential Revision: [D30334815](https://our.internmc.facebook.com/intern/diff/D30334815)

[ghstack-poisoned]

alanwaketan commented:
Looks like I somehow messed up with ghstack. So another #63345 PR is created. Let's use the new PR then.

wconstab reviewed (commented):
LGTM! Just make sure CI works first..
